sns_ik_lib/src/sns_ik.cpp

```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+    if (i < numJnt - 1) {
+      p_0_i_1_i.col(i) = p_0_i_1.col(i+1) - p_0_i_1.col(i);
+    }
+    else {
+      p_0_i_1_i.col(i) = p_0_k - p_0_i_1.col(i);
+    }
+
+    // compute its derivative
+    Eigen::Vector3d p_0_i_1_i_ = p_0_i_1_i.col(i);
+    pdot_0_i_1_i.col(i) = w_i.cross(p_0_i_1_i_);
+
+    // compute the position vector between joint coordinate frame and tip
+    p_0_i_1_k.col(i) = p_0_k - p_0_i_1.col(i);
+  }
+
+  // second pass -- compute Jdv
+  for (size_t i = 0; i < numJnt; i++) {
+    // sum of derivative
+    for (size_t j = i; j < numJnt; j++) {
+      pdot_0_i_1_k.col(i) = pdot_0_i_1_k.col(i) + pdot_0_i_1_i.col(j);
+    }
+    // compute Jdv
+    Eigen::Vector3d Jw_i = Jw.col(i);
+    Eigen::Vector3d Jdw_i = Jdw.col(i);
+    Eigen::Vector3d p_0_i_1_k_ = p_0_i_1_k.col(i);
+    Eigen::Vector3d pdot_0_i_1_k_ = pdot_0_i_1_k.col(i);
+    Jdv.col(i) = Jdw_i.cross(p_0_i_1_k_) + Jw_i.cross(pdot_0_i_1_k_);
+  }
+
+  // assign Jdv and Jdw into Jacobian dot
+  jacobianDot.data.block(0, 0, 3, numJnt) = Jdv;
+  jacobianDot.data.block(3, 0, 3, numJnt) = Jdw;
+
+  // Get the jacobian
+  *jacobianDotOut = jacobianDot.data;
+
+  return true;
+}
+
+int SNS_IK::CartToJntAcc(const KDL::JntArray& q_in, const KDL::JntArray& qdot_in,
+                         const KDL::Twist& a_in,
+                         const KDL::JntArray& q_bias,
+                         const std::vector<std::string>& biasNames,
+                         const KDL::JntArray& q_acc_bias,
+                         KDL::JntArray& qddot_out)
+{
+  if (!m_initialized) {
+    ROS_ERROR("SNS_IK was not properly initialized with a valid chain or limits.");
+    return -1;
+  }
+
+  // compute Jacobian
+  KDL::Jacobian jacobian;
+  jacobian.resize(q_in.rows());
+  if (m_jacobianSolver->JntToJac(q_in, jacobian) < 0)
+  {
+    ROS_ERROR("JntToJac failed!");
+    return -1;
+  }
+
+  // compute Jacobian dot
+  Eigen::MatrixXd jacobianDot;
+  if (!getJacobianDot(q_in, qdot_in, &jacobianDot))
+  {
+    ROS_ERROR("Jacobian dot computation failed!");
+    return -1;
+  }
+
+  std::vector<TaskAcc> sot;
+  TaskAcc task;
+  task.jacobian = jacobian.data;
+  Eigen::VectorXd dJdq = jacobianDot*qdot_in.data;
+  task.dJdq = dJdq;
+  task.desired = Eigen::VectorXd::Zero(6);
+
+  // twistEigenToKDL
+  for(size_t i = 0; i < 6; i++)
+      task.desired(i) = a_in[i];
+  sot.push_back(task);
+
+  // Calculate the nullspace goal as a configuration-space task.
+  // Creates a task Jacobian which maps the provided nullspace joints to
+  // the full joint state.
+  if (q_bias.rows()) {
+    TaskAcc task2;
+    std::vector<int> indicies;
+    if (!nullspaceBiasTask(q_bias, biasNames, &(task2.jacobian), &indicies)) {
+      ROS_ERROR("Could not create nullspace bias task");
+      return -1;
+    }
+    task2.desired = Eigen::VectorXd::Zero(q_bias.rows());
+    for (size_t ii = 0; ii < q_bias.rows(); ++ii) {
+      // This calculates a "nullspace velocity".
```

Comment:
Yes, Good point. The name could now be changed to nullspace acceleration although the calculation stays the same since the desired acceleration is zero.